BUG: Title sometimes is bytes and not str.

reformy opened the pull request:
closes #2929

stefan6419846 commented:
Thanks for the PR. I assume that you own all the necessary copyrights for the test file?

For me this mostly looks okay, apart from the review comment left above about the position after the comment. Nevertheless, I am going to wait for @pubpub-zz here to have another look.

pubpub-zz commented:
Personnaly I would have prefer to add the __str__ to the BytesStringObject() with trying diffrerent encoding(see §3.8.1) .

we should use CHARSET (https://github.com/py-pdf/pypdf/blob/98aa9742e757ec428e1953ba7f47c6d7c44b331a/pypdf/generic/_base.py#L841) adding as first format UTF-16 (based on the spec)

It would then just require to add a call to str in the diffrerent properties of metadata in https://github.com/py-pdf/pypdf/blob/main/pypdf/_doc_common.py

reformy commented:
Thank you for your comments.
I've moved the utf-8 to be after the pdfdocencoding.
@pubpub-zz I did not understand your comment. If the whole str-decoding is not in the right place, that a different issue. I've just added another option to decode it.

stefan6419846 commented:
> If the whole str-decoding is not in the right place, that a different issue.

No, this is related to this issue. The suggestion is to revert the current changes and provide `ByteStringObject.__str__()` accordingly, which would then test the different charsets. At the same time, `DocumentInformation._get_text()` should be rewritten to check for `StringObject` and `ByteStringObject` and - instead of returning the pure value - return `str(retval)`.

reformy commented:
I've changed the code as suggested, adding `__str__` and calling it from `_get_text`. This works well.

However, adding `utf-16` to the `CHARSETS` list breaks other things. I don't understand why do we need this and how is this related to my fix.

stefan6419846 commented on pypdf/generic/_utils.py:
```diff
@@ -189,6 +189,15 @@ def create_string_object(
                 retval.autodetect_pdfdocencoding = True
                 return retval
             except UnicodeDecodeError:
+                # Try utf-8.
```
Isn't this unnecessary now?

stefan6419846 commented:
> However, adding utf-16 to the CHARSETS list breaks other things. I don't understand why do we need this and how is this related to my fix.

This is another encoding to consider. You should only consider this in `__str__`, not add it to the list permanently.

stefan6419846 commented on pypdf/generic/_base.py:
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
+        for enc in NameObject.CHARSETS:
+            try:
+                ret = self.decode(enc)
+                return NameObject(ret)
```
Why do we need to wrap this into a `NameObject`?

stefan6419846 commented on pypdf/generic/_base.py:
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
+            try:
+                ret = self.decode(enc)
+                return NameObject(ret)
+            except Exception:
```
Which actual exceptions can occur here? `Exception` is rather broad and I would have expected to see only `UnicodeDecodeError` here.

stefan6419846 commented on pypdf/generic/_base.py:
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
+                return self.decode(enc)
+            except UnicodeDecodeError:
+                pass
+        return ""
```
Shouldn't this rather raise an exception? For now, I consider it misleading to return an empty string for non-empty inputs.

reformy replied:
Again, the calling code (`DocumentInformation.title`) assumes that `_get_text` will return `''` if failed to decode, and I didn't want to change any unrelated behavior other than the bug I am fixing. But it does make sense, so I will change and hopefully no test will fail.
